src/puppetlabs/puppetdb/command.clj

```diff
@@ -461,60 +456,44 @@
              :response-pub (async/pub response-chan-for-pub :id))))
 
   (start [this context]
-    (let [{:keys [scf-write-db]} (shared-globals)
-          {:keys [response-chan response-pub]} context
-          factory (-> (conf/mq-broker-url (get-config))
-                      (mq/activemq-connection-factory))
-          connection (.createConnection factory)]
+    (let [{:keys [scf-write-db dlo q]} (shared-globals)
+          {:keys [response-chan response-pub]} context]
       (register-listener
        supported-command?
        #(process-command-and-respond! % scf-write-db response-chan (:stats context)))
-      (assoc context
-             :factory factory
-             :connection connection)))
+
+      (upgrade-activemq (get-config)
```

Comment:
foreground seems reasonable to me. I think in 99% of cases the queue will be either small or empty. Putting it in a future adds a bit of complexity and may not buy us much.